File: src/NodeParser/TupleNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { TupleType } from "../Type/TupleType.js";

export class TupleNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.TupleTypeNode): boolean {
        return node.kind === ts.SyntaxKind.TupleType;
    }

    public createType(node: ts.TupleTypeNode, context: Context): BaseType {
        return new TupleType(
            node.elements.map((item) => {
                return this.childNodeParser.createType(item, context);
            }),
        );
    }
}

File: internal/parser/tuple_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// TupleNodeParser parses tuple type nodes (src/NodeParser/TupleNodeParser.ts).
type TupleNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewTupleNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *TupleNodeParser {
	return &TupleNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *TupleNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindTupleType
}

func (p *TupleNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	elements := node.AsTupleTypeNode().Elements.Nodes
	memberTypes := make([]types.Type, len(elements))
	for i, item := range elements {
		memberTypes[i] = p.childNodeParser.CreateType(item, ctx, nil)
	}
	return types.NewTupleType(memberTypes)
}
